src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/OutputProducer.cs

```diff
@@ -112,8 +113,12 @@ private Task FlushAsyncAwaited(WritableBufferAwaitable awaitable)
                     });
                 }
             }
+            await _flushTcs.Task;
 
-            return _flushTcs.Task;
+            if (cancellationToken.IsCancellationRequested)
+            {
+                throw new TaskCanceledException();
```

Comment:
Just call `cancellationToken.ThrowIfCancellationRequested()` afterwards. Maybe this should be:

```C#
if (cancellationToken.CanBeCancelled)
{
    cancellationToken.ThrowIfCancellationRequested();
}
```